src/main/scala/za/co/absa/commons/lang/CollectionImplicits.scala

```diff
@@ -41,4 +46,38 @@ object CollectionImplicits {
     }
   }
 
+  implicit class TraversableOnceOps[A, M[X] <: TraversableOnce[X]](val xs: M[A]) extends AnyVal {
+
+    /**
+      * Almost like `distinct`, but instead of comparing elements itself it compares their projections,
+      * returned by a given function `f`.
+      *
+      * It's logically equivalent to doing stable grouping by `f` followed by selecting the first value of each key.
+      *
+      * @param f   projection function
+      * @param cbf collection builder factory
+      * @return a new sequence of elements which projection (obtained by applying the function `f`)
+      *         are the first occurrence of every other elements' projection of this sequence.
+      */
+    def distinctBy[B](f: A => B)(implicit cbf: CanBuildFrom[M[A], A, M[A]]): M[A] = {
+      val seen = mutable.Set.empty[B]
+      val b = cbf(xs)
+      for (x <- xs) {
+        val y = f(x)
+        if (!seen(y)) {
+          b += x
+          seen += y
+        }
+      }
+      b.result()
```

Comment:
There seems to be an abundance of single-letter variables which does not always seem the best in terms of readability e.g. `y` could have been named `projection` etc. But it makes sense overall, just an observation.

The logic seems correct and efficient to me.

Reply:
Thanks for pointing that out. That was my doubts as well. The reason they named like that is basically because the code itself is basically a copy-paste from `SeqLike .distinct` method. So I just wanted to retain the visual similarity with the original, for whatever reason :) The variable `y` was the only addition, and was named after `y = f(x)`, following the existing style.

Reply:
Sure, I understand, such a perspective is absolutely appropriate. As I said just an observation, don't worry about it too much :D